### ISSUE TITLE
Implement the ICCV 2015 best paper "Deep Neural Decision Forests"

### ISSUE THREAD

**Issue by futurely** [dmlc/mxnet#1083]
The creators of Deep Neural Decision Forests [1] implemented their networks in DMLC/CXXNET which has been superseded by this project. It is much more efficient to estimate the leaf node probability distributions with a large mini-batch instead of the whole training set. One of the major advantages of CXXNET brought to their proposed dNDF.NET was distributed training in which effective mini-batch size was the single node mini-batch size multiplied with the number of cluster nodes and at the same time training was much faster than on a single node.

By far, the only attempt at implementing the paper on GitHub is [based on Theano](https://github.com/diogo149/treeano/blob/master/treeano/theano_extensions/tree_probability.py) and [incomplete](https://github.com/diogo149/treeano/blob/master/treeano/sandbox/nodes/dNDF.py).

[1] P. Kontschieder, M. Fiterau, A. Criminisi, and S. Rota Bulo'. [Deep Neural Decision Forests](http://research.microsoft.com/apps/pubs/default.aspx?id=255952). ICCV 2015.

**Comment by SkidanovAlex** [comment]
I built neural decision forests in Lasagne (https://github.com/SkidanovAlex/ShallowNeuralDecisionForest), where I actually managed to evaluate leaf probabilities in a vector form, allowing the entire model to be executed on a GPU. I was, however, never able to get as good results as theirs (it was on par with a regular NN on all the tasks I tried it, but was taking longer to converge). They never responded to my email to them with request for more details, so I don't know what exactly I am doing different.
